cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -461,6 +463,224 @@ class ObjectInputFile final : public io::RandomAccessFile {
+              "GetProperties failed for '" + block_blob_client_->GetUrl() +
+                  "' with an unexpected Azure error. Can not initialise an "
+                  "ObjectAppendStream without knowing whether a file already exists at "
+                  "this path, and if it exists, its size.",
+              exception);
+        }
+        content_length_ = 0;
+      }
+    }
+    if (content_length_ > 0) {
+      ARROW_ASSIGN_OR_RAISE(auto block_list, GetBlockList(block_blob_client_));
+      for (auto block : block_list.CommittedBlocks) {
+        block_ids_.push_back(block.Name);
+      }
+    }
+    return Status::OK();
+  }
+
+  Status Abort() override {
+    if (closed_) {
+      return Status::OK();
+    }
+    block_blob_client_ = nullptr;
+    closed_ = true;
+    return Status::OK();
+  }
+
+  Status Close() override {
+    if (closed_) {
+      return Status::OK();
+    }
+    RETURN_NOT_OK(Flush());
+    block_blob_client_ = nullptr;
+    closed_ = true;
+    return Status::OK();
+  }
+
+  bool closed() const override { return closed_; }
+
+  Status CheckClosed(const char* action) const {
+    if (closed_) {
+      return Status::Invalid("Cannot ", action, " on closed stream.");
+    }
+    return Status::OK();
+  }
+
+  Result<int64_t> Tell() const override {
+    RETURN_NOT_OK(CheckClosed("tell"));
+    return pos_;
+  }
+
+  Status Write(const std::shared_ptr<Buffer>& buffer) override {
+    return DoAppend(buffer->data(), buffer->size(), buffer);
+  }
+
+  Status Write(const void* data, int64_t nbytes) override {
+    return DoAppend(data, nbytes);
+  }
+
+  Status DoAppend(const void* data, int64_t nbytes,
+                  std::shared_ptr<Buffer> owned_buffer = nullptr) {
+    RETURN_NOT_OK(CheckClosed("append"));
+    auto append_data = reinterpret_cast<uint8_t*>((void*)data);
+    auto block_content = Azure::Core::IO::MemoryBodyStream(
+        append_data, strlen(reinterpret_cast<char*>(append_data)));
+    if (block_content.Length() == 0) {
+      return Status::OK();
+    }
+
+    auto size = block_ids_.size();
+
+    // New block ids must always be distinct from the existing block ids. Otherwise we
+    // will accidentally replace the content of existing blocks, causing corruption.
+    // We will use monotonically increasing integers.
+    std::string new_block_id = std::to_string(size);
+
+    // Pad to 5 digits, because Azure allows a maximum of 50,000 blocks.
+    const size_t target_number_of_digits = 5;
+    int required_padding_digits =
+        target_number_of_digits - std::min(target_number_of_digits, new_block_id.size());
+    new_block_id.insert(0, required_padding_digits, '0');
+    new_block_id += "-arrow";  // Add a suffix to reduce risk of block_id collisions with
+                               // blocks created by other applications.
+    new_block_id = Azure::Core::Convert::Base64Encode(
+        std::vector<uint8_t>(new_block_id.begin(), new_block_id.end()));
+
```

Comment:
If we want to be 100% confident of avoiding clashes then yes but personally I think the current solution is a good compromise. 

The risk should be zero when using `OpenOutputStream` because every block ID will be created by this same scheme, using monotonically increasing integers. The risk when using `OpenAppendStream` is that previously committed blocks used unusual names that might conflict. For example if some other writer committed block blob consisting of one block named `00002-arrow` then that would conflict after this writer appends 2 additional blocks, and cause a corrupt blob. I think this is extremely unlikely so personally I think this is a good option. Additionally `OpenAppendStream` is not implemented at all for S3 and GCS so presumably its not used much.

Reply:
OK. Could you describe the risk as a comment? If we find a real world problem with the risk, we can revisit the risk.

Reply:
Done